File: implement-shell-tools/wc/wc.js
import { readFile } from "node:fs/promises";
import process from "node:process";

const argvs= process.argv.slice(2);
let countLines = false;
let countWords = false;
let countBytes = false;
let totalWords = 0;
let totalLines = 0;
let totalBytes = 0;
const files = []

for (const arg of argvs) {
  if (arg === "-l") {
    countLines = true;
  } else if (arg === "-w") {
    countWords = true;
  } else if (arg === "-c") {
    countBytes = true;
  } else {
    files.push(arg);
  }
}

for (const file of files) {
  try {
    const content = await readFile(file, "utf-8");
    const arrayOfLines = content.split("\n");
    const arrayOfWords = content.trim().split(/\s+/)
    const bytes = Buffer.byteLength(content, "utf-8");
    const lines = arrayOfLines.length -1;
    const words = arrayOfWords.length;
    totalLines += lines;
    totalBytes += bytes;
    totalWords +=words
    if (countLines && countWords) {
      process.stdout.write(`       ${lines}      ${words} ${file}\n`);

    } else if (countLines) {
      process.stdout.write(`       ${lines} ${file}\n`);
      
    } else if (countWords){
      process.stdout.write(`  ${words} ${file}\n`);

    } else if (countBytes) {
      process.stdout.write(`   ${bytes} ${file}\n`);

    } else {
      process.stdout.write(`       ${lines}      ${words}      ${bytes} ${file}\n`);
    }
      
  } catch(err) {
    console.error(err.message);
  }
}

if (files.length > 1 ) {
  if (countLines && countWords) {
    process.stdout.write(`       ${totalLines}      ${totalWords} total\n`);

  } else if (countLines) {
    process.stdout.write(`       ${totalLines} total\n`);
    
  } else if (countWords){
    process.stdout.write(`  ${totalWords} total\n`);

  } else if (countBytes) {
    process.stdout.write(`   ${totalBytes} total\n`);

  } else {
    process.stdout.write(`       ${totalLines}      ${totalWords}      ${totalBytes} total\n`);
  }
}

File: implement-shell-tools/wc/wc.py
import argparse


parser = argparse.ArgumentParser()

parser.add_argument("-l", action="store_true")
parser.add_argument("-w", action="store_true")
parser.add_argument("-c", action="store_true")
parser.add_argument("files", nargs="+")

args = parser.parse_args()


def add_count(output, value):
  output.append(str(value))


total_lines = 0
total_words = 0
total_bytes = 0

for filename in args.files:

  with open(filename, "rb") as file:
    content = file.read()

  lines = content.count(b"\n")
  words = len(content.split())
  bytes_count = len(content)

  total_lines += lines
  total_words += words
  total_bytes += bytes_count

  if not args.l and not args.w and not args.c:
    print(f"{lines:8} {words:8} {bytes_count:8} {filename}")

  else:
    output = []

    if args.l:
      add_count(output, lines)

    if args.w:
      add_count(output, words)

    if args.c:
      add_count(output, bytes_count)

    print(f"{' '.join(output):>8} {filename}")

